python/python_helper_sympy.py

```diff
@@ -264,98 +323,105 @@ def evaluate(expr, locals_for_eval={}):
-    return s[ind - w_left : ind + w_right] + "\n" + " " * w_left + "^" + " " * w_right
+def point_to_error(expr: str, ind: int, w: int = 5) -> str:
+    """Generate a string with a pointer to error in expr with index ind"""
+    w_left: str = " " * (ind - max(0, ind - w))
+    w_right: str = " " * (min(ind + w, len(expr)) - ind)
+    initial: str = expr[ind - len(w_left) : ind + len(w_right)]
+    return f"{initial}\n{w_left}^{w_right}"
 
 
-def sympy_to_json(a, allow_complex=True):
+def sympy_to_json(
+    a: sympy.Expr, *, allow_complex: bool = True, allow_trig_functions: bool = True
+) -> SympyJson:
     const = _Constants()
 
     # Get list of variables in the sympy expression
-    variables = [str(v) for v in a.free_symbols]
+    variables = list(map(str, a.free_symbols))
 
     # Check that variables do not conflict with reserved names
-    reserved = {
-        **const.helpers,
-        **const.variables,
-        **const.hidden_variables,
-        **const.functions,
-    }
+    reserved = (
+        const.helpers.keys()
+        | const.variables.keys()
+        | const.hidden_variables.keys()
+        | const.functions.keys()
+    )
     if allow_complex:
-        reserved = {
-            **reserved,
-            **const.complex_variables,
-            **const.hidden_complex_variables,
-        }
-    for k in reserved.keys():
-        for v in variables:
-            if k == v:
-                raise ValueError(
-                    "sympy expression has a variable with a reserved name: {:s}".format(
-                        k
-                    )
-                )
+        reserved |= (
+            const.complex_variables.keys() | const.hidden_complex_variables.keys()
+        )
+    if allow_trig_functions:
+        reserved |= const.trig_functions.keys()
+
+    # Check if reserved variables conflict, raise an error if they do
+    conflicting_reserved_variables = reserved & set(variables)
+
+    if conflicting_reserved_variables:
+        raise ValueError(
+            f"sympy expression has a variable with a reserved name: {conflicting_reserved_variables}"
+        )
 
     # Apply substitutions for hidden variables
-    a = a.subs(
-        [(const.hidden_variables[key], key) for key in const.hidden_variables.keys()]
-    )
+    a_sub = a.subs([(val, key) for key, val in const.hidden_variables.items()])
     if allow_complex:
-        a = a.subs(
-            [
-                (const.hidden_complex_variables[key], key)
-                for key in const.hidden_complex_variables.keys()
-            ]
+        a_sub = a_sub.subs(
+            [(val, key) for key, val in const.hidden_complex_variables.items()]
         )
 
-    return {"_type": "sympy", "_value": str(a), "_variables": variables}
+    return {"_type": "sympy", "_value": str(a_sub), "_variables": variables}
 
 
-def json_to_sympy(a, allow_complex=True):
+def json_to_sympy(
+    a: SympyJson, *, allow_complex: bool = True, allow_trig_functions: bool = True
```

Comment:
For my own knowledge, what's the `*` here?

Reply:
It enforces that subsequent keywords must be named, i.e. original allows `json_to_numpy(a, False)` while latter requires `json_to_numpy(a, allow_complex=False)`

Reply:
neat, thanks for explaining!

Reply:
Yep! The `*` is usually used for additional options that get defaulted which don't interact with core data.